adding Overview of Chainguard Enforce (#225)
chainguard-dev/edu: content/chainguard/chainguard-enforce/chainguard-enforce-kubernetes/enforce-overview/index.md

---
title: "Overview of Chainguard Enforce"
type: "article"
description: "A conceptual overview of Chainguard Enforce and its components"
date: 2022-12-15T16:36:47+00:00
lastmod: 2022-12-15T16:36:47+00:00
draft: false
images: []
menu:
  docs:
    parent: "enforce-overview"
weight: 100
toc: true
---

> _This document relates to Chainguard Enforce. You can request access to Chainguard Enforce through selecting **Chainguard Enforce for Kubernetes** on our [inquiry form](https://www.chainguard.dev/get-demo?utm_source=docs)._

An organization's software supply chain can contain many possible points of entry for malicious actors and lots of potential for unintended vulnerabilities. Indeed, with recent high-profile security incidents like the SolarWinds breach and the discovery of the Log4Shell vulnerability, [software supply chain security](/software-security/what-is-software-supply-chain-security/) has become a top priority for many organizations.

> To help keep organizations' software supply chains secure by default, Chainguard has developed **Chainguard Enforce, a supply chain security solution for containerized workloads**. Chainguard Enforce enables you to ensure continuous compliance by enforcing policies that protect organizations from supply chain threats. Using open source projects and standards that are trusted by the community — like Cosign and Fulcio from the Sigstore project — Chainguard Enforce offers a robust approach to securing your workloads, allowing you to do the following:

Reviewer: ```suggestion
**Chainguard Enforce** offers a supply chain security solution for containerized workloads, which can help keep organizations' software supply chains secure by default. Enabling engineering and security teams to ensure continuous compliance, Chainguard Enforce protects organizations from supply chain threats through policy enforcement. Using open source projects and standards that are trusted by the community — like Cosign and Fulcio from the Sigstore project — Chainguard Enforce offers a robust approach to securing your workloads, allowing the following:
```

* Sign any software artifact — including container images, code commits, and more — with an identity-based one-time-use signature
* Meet compliance requirements such as SLSA levels, CIS benchmarks, Pod Security Standards, and more
* Ensure continuous compliance in real time of all of your organization's policies with instant notification of any violations

This article provides an overview of Chainguard Enforce and its components.

## What is Chainguard Enforce?

Chainguard Enforce is a developer platform for end-to-end software supply chain security that operates at the level of Kubernetes fleets. Once you install Enforce onto a Kubernetes cluster and apply a policy, Chainguard Enforce will immediately begin analyzing metadata associated with images running in the cluster to determine whether it satisfies the requirements defined within the policy. Specifically, it inspects the given cluster's [software artifacts](https://console.enforce.dev/policies/catalog), such as an image's software bill of materials (SBOM). Based on the details within the images' metadata, Chainguard Enforce will determine whether those images satisfy the policy.

> Chainguard Enforce provides a rich IAM model similar to the likes of AWS and GCP, allowing you to organize policies and clusters into a hierarchy of groups through its Identity and Access Management (IAM) model. When you install Enforce onto a cluster, you associate that cluster with an existing IAM group. Likewise, when you create a policy, you associate it with a group and it is automatically applied to any clusters already associated with that group. 

Reviewer: ```suggestion
Chainguard Enforce provides a rich [IAM model](../overview-of-enforce-iam-model/) similar to the likes of AWS and GCP, allowing you to organize policies and clusters into a hierarchy of groups through its Identity and Access Management (IAM) model. When you install Enforce onto a cluster, you associate that cluster with an existing IAM group. Likewise, when you create a policy, you associate it with a group and it is automatically applied to any clusters already associated with that group. 
```

> One feature that makes Chainguard Enforce unique is continuous verification. This means that Enforce will scan the cluster on a regular cadence to ensure that it complies with any policies you've applied to it. Chainguard Enforce uses an [admission controller](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/) to block any requests that violate a cluster's policy. 

Reviewer: ```suggestion
One feature that makes Chainguard Enforce unique is [continuous verification](../understanding-continuous-verification/). This means that Enforce will scan the cluster on a regular cadence to ensure that it complies with any policies you've applied to it. Chainguard Enforce uses an [admission controller](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/) to block any requests that violate a cluster's policy. 
```

Typically, admission controllers only check policies when Kubernetes resources are being created. If new policies are rolled out, or the state of an application’s compliance degrades over time, then admission controllers will not catch the compliance degradation. Chainguard Enforce, however, continuously scans the cluster against all of the latest policies. This means that if a deployed application violates a new policy or falls out of compliance, it will catch the policy violation before Kubernetes acts on the request. 

Violations can be sent as Cloud Events, so you can respond in the most appropriate way for your policies and applications. This might be to file a JIRA ticket, page the on-call, delete pods, or trigger a CI/CD pipeline to redeploy the application. Regardless of how you configure your response, the violation will be logged for future reference.

> > Note: For more information on how continuous verification works, check out our conceptual article on [Understanding Continuous Verification](../understanding-continuous-verification/).

Reviewer: I would remove this note and put the link in line

The following diagram outlines Chainguard Enforce's various components and how they work together. 

> ![Diagram outlining Chainguard Enforce's architecture.](enforce-diagram.png)

Reviewer: On the dark background the image is hard to read:

<img width="995" alt="Screenshot 2022-12-15 at 4 48 47 PM" src="https://user-images.githubusercontent.com/5977693/207974441-1ba32dad-28b2-4eda-8586-b92171ae98f0.png">

Author: I added a white background and bumped up the font size across the board.

Enforce can also be configured to send events to external systems like Slack, JIRA, or GitHub to notify users of policy violations, configuration changes, and more.  Finally, Enforce offers a command line interface, `chainctl`, for users to interact with clusters and policies from the command line.

## Managing Chainguard Enforce with chainctl

`chainctl` — short for "Chainguard Control" — is the command line interface tool for interacting with Chainguard Enforce. It communicates with the Enforce API, allowing you to manage policies, IAM groups, and cluster connections. Assuming you've been granted access to the product, you can [install `chainctl` for Chainguard Enforce](/chainguard/chainguard-enforce/how-to-install-chainctl/) on any machine you use to manage your Kubernetes clusters.

`chainctl` uses the familiar `<context> <noun> <verb>` style of CLI interactions. For example, to list groups within the context of Chainguard IAM, you can run `chainctl iam groups list` to receive relevant output.

For more information on how to use `chainctl`, please refer to our [`chainctl` documentation](/chainguard/chainguard-enforce/chainctl-docs/chainctl/).

> ## The Chainguard Enforce Agent

Reviewer: Do you think a page on the Enforce Agent that is a little more detailed is warranted?

Author: warranted, yes. But probably not a necessity right now

> In the context of computing, a software agent is a program that performs certain actions on behalf of a user. A software agent has the authority to make decisions on its own and doesn't necessarily require user input. Accordingly, the Chainguard Enforce Agent is able to analyze a given cluster and enforce any policies you've applied to the cluster without your oversight. 

Reviewer: ```suggestion
In the context of computing, a software agent is a program that performs certain actions on behalf of a user. A software agent has the authority to make decisions on its own and doesn't necessarily require user input. Accordingly, the Chainguard Enforce Agent is able to analyze a given cluster and enforce any policies you've applied to the cluster without manual human oversight. 
```

To better understand how policies work in Chainguard Enforce, we encourage you to review the [Sigstore Policy Controller documentation](https://docs.sigstore.dev/policy-controller/overview). In particular, you may want to review the Policy Controller [Admission of images documentation](https://docs.sigstore.dev/policy-controller/overview/#admission-of-images) to understand how to admit images through the cluster image policy. Additionally, Chainguard has assembled a [policy catalog](https://console.enforce.dev/policies/catalog) in the Chainguard Enforce Console which provides a number of sample policies you can use or edit to suit your own needs.

## Chainguard Enforce for Git

> In addition to ensuring that Kubernetes clusters comply with established policies, Chainguard Enforce also supports Git signature verification for GitHub repositories. Chainguard Enforce for Git is built around [Gitsign](https://docs.sigstore.dev/gitsign/overview/) a component of the [Sigstore project](https://www.sigstore.dev/). Gitsign implements keyless Sigstore to sign Git commits with a valid OpenID Connect identity, like a GitHub or Google account. This allows users to sign their Git commits without the need for GPG keys or a complicated configuration.

Reviewer: ```suggestion
In addition to ensuring that Kubernetes clusters comply with established policies, Chainguard Enforce also supports Git signature verification for GitHub repositories. Chainguard Enforce for Git is built around [Gitsign](https://docs.sigstore.dev/gitsign/overview/) a tool that is part of the [Sigstore project](https://www.sigstore.dev/). Gitsign implements keyless Sigstore to sign Git commits with a valid OpenID Connect identity, like a GitHub or Google account. This allows users to sign their Git commits without the need for GPG keys or a complicated configuration.
```

Reviewer: This is a nit, just that it is a tool on its own as part of the Sigstore "suite"

Author: that's very helpful, I was having a hard time describing it!
